File: app/lib/network/probe_priority.dart
import '../api/api.dart';

/// Probe tier for reachability scans.
enum ProbePriority {
  /// LAN-discovered or has a direct HTTP URL.
  lanDiscovered,

  /// Same-account device with server presence online.
  presenceOnline,

  /// Lazy: skip auto-probe; show offline until user selects or manual refresh.
  lazy,
}

class ProbePartition {
  const ProbePartition({
    required this.lanDiscovered,
    required this.presenceOnline,
    required this.lazy,
  });

  final List<DeviceDto> lanDiscovered;
  final List<DeviceDto> presenceOnline;
  final List<DeviceDto> lazy;

  List<DeviceDto> get autoProbe => [...lanDiscovered, ...presenceOnline];
}

ProbePriority classifyDevice(
  DeviceDto device, {
  required Set<String> nearbyIds,
  required Set<String> myDeviceIds,
}) {
  final hasLanUrl =
      device.lanHttpUrl != null && device.lanHttpUrl!.trim().isNotEmpty;
  if (hasLanUrl || nearbyIds.contains(device.deviceId)) {
    return ProbePriority.lanDiscovered;
  }
  if (myDeviceIds.contains(device.deviceId) &&
      device.presenceStatus == 'online') {
    return ProbePriority.presenceOnline;
  }
  return ProbePriority.lazy;
}

ProbePartition partitionForProbe(
  List<DeviceDto> devices, {
  required Set<String> nearbyIds,
  required Set<String> myDeviceIds,
}) {
  final p0 = <DeviceDto>[];
  final p1 = <DeviceDto>[];
  final p2 = <DeviceDto>[];
  for (final d in devices) {
    switch (classifyDevice(
      d,
      nearbyIds: nearbyIds,
      myDeviceIds: myDeviceIds,
    )) {
      case ProbePriority.lanDiscovered:
        p0.add(d);
      case ProbePriority.presenceOnline:
        p1.add(d);
      case ProbePriority.lazy:
        p2.add(d);
    }
  }
  return ProbePartition(
    lanDiscovered: p0,
    presenceOnline: p1,
    lazy: p2,
  );
}

/// Returns true when an auto-probe can be skipped (offline with no LAN URL).
bool shouldSkipAutoProbe(
  DeviceDto device, {
  required Set<String> nearbyIds,
}) {
  final hasLanUrl =
      device.lanHttpUrl != null && device.lanHttpUrl!.trim().isNotEmpty;
  if (hasLanUrl || nearbyIds.contains(device.deviceId)) return false;
  return device.presenceStatus == 'offline';
}

File: web/src/lib/probePriority.ts
import type { DeviceDto } from '@/lib/api';

export type ProbePriority = 'lanDiscovered' | 'presenceOnline' | 'lazy';

export type ProbePartition = {
  lanDiscovered: DeviceDto[];
  presenceOnline: DeviceDto[];
  lazy: DeviceDto[];
};

export function classifyDevice(
  device: DeviceDto,
  nearbyIds: Set<string>,
  myDeviceIds: Set<string>,
): ProbePriority {
  const hasLanUrl = !!(device.lanHttpUrl && device.lanHttpUrl.trim().length > 0);
  if (hasLanUrl || nearbyIds.has(device.deviceId)) return 'lanDiscovered';
  if (myDeviceIds.has(device.deviceId) && device.presenceStatus === 'online') {
    return 'presenceOnline';
  }
  return 'lazy';
}

export function partitionForProbe(
  devices: DeviceDto[],
  nearbyIds: Set<string>,
  myDeviceIds: Set<string>,
): ProbePartition {
  const lanDiscovered: DeviceDto[] = [];
  const presenceOnline: DeviceDto[] = [];
  const lazy: DeviceDto[] = [];
  for (const d of devices) {
    switch (classifyDevice(d, nearbyIds, myDeviceIds)) {
      case 'lanDiscovered':
        lanDiscovered.push(d);
        break;
      case 'presenceOnline':
        presenceOnline.push(d);
        break;
      case 'lazy':
        lazy.push(d);
        break;
    }
  }
  return { lanDiscovered, presenceOnline, lazy };
}

export function shouldSkipAutoProbe(device: DeviceDto, nearbyIds: Set<string>): boolean {
  const hasLanUrl = !!(device.lanHttpUrl && device.lanHttpUrl.trim().length > 0);
  if (hasLanUrl || nearbyIds.has(device.deviceId)) return false;
  return device.presenceStatus === 'offline';
}

export async function runWithConcurrency<T>(
  items: T[],
  concurrency: number,
  fn: (item: T) => Promise<void>,
): Promise<void> {
  if (items.length === 0) return;
  const limit = Math.max(1, Math.min(concurrency, items.length));
  let nextIndex = 0;
  async function worker() {
    while (nextIndex < items.length) {
      const index = nextIndex++;
      await fn(items[index]);
    }
  }
  await Promise.all(Array.from({ length: limit }, () => worker()));
}
